### PR TITLE
Issue/2836

### PR DESCRIPTION
**Pull request by juazasan**
# Description

Changes to tracing libraries to leverage OpenTelemetry instead of Opencensus.
Added support for otel protocol in addition to zipkin to allow exporting traces to an opentelemetry agent using native protocol over both http and grpc.

## Issue reference

Please reference the issue this PR will close: #2836 

## Checklist

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Unit tests passing
* [X] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/doc#2405

### CODE REVIEW

**Comment by juazasan** [comment]
@yaron2 Please, need some specific feedback on how I dealt with the fact Spans in Opentelemetry are an interface instead of a struct like in Opencesus. It caused some issues with the existing functions when tried to pass variables as references to an interface. I'm not sure if the workaround I applied follows golang's best practices and whether I should have better used a wrapper. Please let me know your thoughts. Thanks.

File: pkg/channel/http/http_channel.go [hunk 1]
```diff
@@ -222,8 +221,10 @@ func (h *Channel) constructRequest(ctx context.Context, req *invokev1.InvokeMeth
 
 	// HTTP client needs to inject traceparent header for proper tracing stack.
 	span := diag_utils.SpanFromContext(ctx)
-	httpFormat := &tracecontext.HTTPFormat{}
-	tp, ts := httpFormat.SpanContextToHeaders(span.SpanContext())
+	// httpFormat := &tracecontext.HTTPFormat{}
+	// tp, ts := httpFormat.SpanContextToHeaders((*span).SpanContext())
```
Comment (yaron2) [inline pkg/channel/http/http_channel.go:225]: Remove commented code please

**Comment by yaron2** [comment]
@juazasan please resolve the conflicts. I'll then test this branch to make a sanity check.

**Review by yaron2 — CHANGES_REQUESTED** [review]

File: pkg/channel/http/http_channel.go (hunk 1 above)
Comment (juazasan) [reply pkg/channel/http/http_channel.go]: Done

**Comment by juazasan** [comment]
> @juazasan please resolve the conflicts. I'll then test this branch to make a sanity check.

I am working on it. It seems one of the e2e test recently added is broken. Will need some time to debug the test case and see what is the problem.

**Comment by ItalyPaleAle** [comment]
@juazasan Our E2E tests are a bit flaky (we're working on it... but it's not an easy task :) ). But I see that they're passing in most cases and the single one failing seems to be due to unrelated reasons. So I wouldn't worry about that.

I also just triggered the E2E tests in our other environment (running on a real Kubernetes cluster on Azure) to see if they work. You'll get the response here in about 1hr.

However, @juazasan, I see that the [DCO check is failing](https://github.com/dapr/dapr/pull/4781/checks?check_run_id=6960877545). You will need to fix that, please!

**Comment by ItalyPaleAle** [comment]
Just need to edit the commit message for the commits that are failing the DCO test (see link above). You don't really need to rebase anything - the `git rebase` command just happens to be the easiest one to use to change a commit message.

**Comment by juazasan** [comment]
> Just need to edit the commit message for the commits that are failing the DCO test (see link above). You don't really need to rebase anything - the `git rebase` command just happens to be the easiest one to use to change a commit message.

Please, do you have a link to something that explains how to do that? Tried a couple of things but they did not work. Thanks

**Comment by ItalyPaleAle** [comment]
You can read the instructions here: https://github.com/dapr/dapr/pull/4781/checks?check_run_id=6960877545

You need to run `git rebase HEAD~9 --signoff` then `git push --force`.

**Comment by juazasan** [comment]
> You can read the instructions here: https://github.com/dapr/dapr/pull/4781/checks?check_run_id=6960877545
> 
> You need to run `git rebase HEAD~9 --signoff` then `git push --force`.

@ItalyPaleAle Tried my best to fix the above but ended up entering an endless process of fixing conflicts for each commit. I am starting a new PR where I am just adding all the changes on top of master. Hope that works.

**Pull request closed without merging**